CONTRIBUTING_GO.md

```diff
@@ -161,6 +164,27 @@ reasons why this is _the right thing to do_:
 License: MIT
 Signed-off-by: Juan Benet <juan@ipfs.io>
 ```
+
+## Contributor Principles
+
+These principles are a living document and should be updated as-needed to better serve the go-ipfs community.
+
+In order of precedence:
+
+1. **Build the distributed web.** (may need a better way to capture this)
+2. **Don't break userspace.** This isn't "don't break protocols"; we do, in fact, occasionally break protocols (with care) when we know that users won't notice it.
```

Comment:
Let's clarify what this *IS* then:
```suggestion
2. **Don't break userspace.** This isn't "don't break protocols"; we do, in fact, occasionally break protocols (with care) when we know that users won't notice it. This means support our users' varieties of needs - past, present, and future!
```

Reply:
I'm going more for the Linux version of "don't break userspace". That is:

1. Don't change stable APIs that people are relying on.
2. Don't break workflows (within reason).
3. Don't break compatibility (within reason).

Reply:
I guess this point is really about building trust.

Reply:
Cool - let's add that then?

Reply:
FWIW, I've never heard the term "userspace" before. Is there a way to describe this concept with different words that are more accessible? From context I guessed it meant "Don't break the user experience" or "Don't make changes that will break apps people already built on top of our code," but Google says it's two words (user space) and means "code that runs outside the operating system's kernel."

Reply:
I've seen "userspace" and "ecosystem" used both as differentiated terms and as essentially the same term, so their use without some qualifier is a bit confusing.

Often, "ecosystem" refers to the things built on top of the API/platform directly and "userspace" refers to the users and applications built on that ecosystem.

In Linux, they'll break kernel modules with the understanding that those modules will be updated before release and therefor will not break "userspace" but **will** break ecosystem.

In Node.js, the project can't really break the ecosystem because old packages that depend on those APIs will live in on as deep dependencies in the userspace indefinitely.

Basically, you're allowed to break things provided that you can bring along enough of the ecosystem with you in a reasonable time frame. The way to differentiate "acceptable breakage" from "unacceptable breakage" is the degree to which you think the userspace will be effected.

Reply:
> FWIW, I've never heard the term "userspace" before. Is there a way to describe this concept with different words that are more accessible? From context I guessed it meant "Don't break the user experience" or "Don't make changes that will break apps people already built on top of our code," but Google says it's two words (user space) and means "code that runs outside the operating system's kernel."

This comes from the Linux kernel development and yes, "userspace" isn't really the right term because we're not building a kernel. However, I'd like to find some way to draw the parallel to the Linux kernel because many go-ipfs contributors will be familiar with that phrase and what it entails.

> In Node.js, the project can't really break the ecosystem because old packages that depend on those APIs will live in on as deep dependencies in the userspace indefinitely.

This is more what we're looking for here, I'll try to find a nice way to put this.